src/Altinn.App.Core/Expressions/LayoutEvaluatorState.cs

```diff
@@ -0,0 +1,199 @@
+using System;
+using System.Text.Json;
+using Altinn.App.Core.Configuration;
+using Altinn.App.Core.Helpers;
+using Altinn.App.Core.Models.Expressions;
+using Altinn.App.Core.Models.Layout;
+using Altinn.Platform.Storage.Interface.Models;
+
+namespace Altinn.App.Core.Expressions;
+
+/// <summary>
+/// Collection class to hold all the shared state that is required for evaluating expressions in a layout.
+/// </summary>
+public class LayoutEvaluatorState
+{
+    private readonly IDataModelAccessor _dataModel;
+    private readonly LayoutModel _componentModel;
+    private readonly FrontEndSettings _frontEndSettings;
+    private readonly Instance _instanceContext;
+
+    /// <summary>
+    /// Constructor for LayoutEvaluatorState. Usually called via <see cref="LayoutEvaluatorStateInitializer" /> that can be fetched from dependency injection.
+    /// </summary>
+    public LayoutEvaluatorState(IDataModelAccessor dataModel, LayoutModel componentModel, FrontEndSettings frontEndSettings, Instance instance)
+    {
+        _dataModel = dataModel;
+        _componentModel = componentModel;
+        _frontEndSettings = frontEndSettings;
+        _instanceContext = instance;
+        // TODO: shaddowFields ...
+    }
+
+
+    /// <summary>
+    /// Get a hierarcy of the different contexts in the component model (remember to iterate <see cref="ComponentContext.ChildContexts" />)
+    /// </summary>
+    public IEnumerable<ComponentContext> GetComponentContexts()
+    {
+        if (_componentModel is null || _dataModel is null)
```

Comment:
I think these checks and exception throwing should go in the constructor to ensure the object is initialized in a valid state instead of deferring the check to when a method is called.

Reply:
I just dropped the null checks. They are declared as non nullable in the constructor, so the compiler will check, or if you disable nullable reference types, you deserve a null pointer exception.

The reason I did the check previously, was that I wanted the properties to be nullable, and have some way of figuring out what data is actually needed for a given evaluation. It turned out to be too complex, so I just gave up.
085b143

Reply:
> I’m not completely convinced, at least not if you think this should be marked `internal` instead of `public`. Seems like this would be really useful to have available in apps and other code, and because of the json in apps, it needs to be very stable.

No, I don't mean the classes should be internal - just moved to the internal namespace together with the other features that is internal. The reasoning is partly to avoid all internal features on root but mostly to signal that this is internal stuff, and if you replace implementations from this namespace then you need to be sure of what you are doing.